### ISSUE TITLE
`torch.scatter_reduce` CPU performance regression (~15–58×) introduced in 2.13.0.

### ISSUE THREAD

**Issue by seanonymoussubmission** [pytorch/pytorch#183812]
### 🐛 Describe the bug

`torch.scatter_reduce` on CPU is 15–58× slower in the `2.13.0.dev20260512+cu130` nightly compared to `20260428` and `20260505`. GPU time is identical across all three builds, isolating the regression to the CPU dispatch or algorithm path.
Two shapes and reduce ops are affected. Both were fine on Apr 28 and May 5 and first regressed on May 12.

## To Reproduce

```python
import torch
import time
import statistics

def bench_cpu(fn, *args, n=30, **kwargs):
    fn(*args, **kwargs)  # warmup
    samples = [0.0] * n
    for i in range(n):
        t0 = time.perf_counter()
        fn(*args, **kwargs)
        samples[i] = time.perf_counter() - t0
    return statistics.median(samples) * 1e3  # ms

rng = torch.Generator()
rng.manual_seed(880014)

# Shape A: float32 amax
input_a = torch.randn(128, 64, 128, dtype=torch.float32)
index_a = torch.randint(0, 64, (128, 256, 128), dtype=torch.int64)
src_a   = torch.randn(128, 256, 128, dtype=torch.float32)
t_a = bench_cpu(torch.scatter_reduce,
                input_a.clone(), 1, index_a, src_a,
                reduce="amax", include_self=False)
print(f"Shape A: {t_a:.2f} ms")   # expect ~1.6 ms; regressed to ~24 ms on May 12

# Shape B: float16 mean
input_b = torch.randn(32, 128, 512, dtype=torch.float16)
index_b = torch.randint(0, 128, (32, 256, 512), dtype=torch.int64)
src_b   = torch.randn(32, 256, 512, dtype=torch.float16)
t_b = bench_cpu(torch.scatter_reduce,
                input_b.clone(), 1, index_b, src_b,
                reduce="mean", include_self=True)
print(f"Shape B: {t_b:.2f} ms")   # expect ~1.3 ms; regressed to ~78 ms on May 12
```

---

## Expected Behavior

CPU median latency consistent with the Apr 28 baseline:

| Shape | dtype | reduce | include_self | Expected |
|-------|-------|--------|--------------|----------|
| `[128, 64, 128]` input, `[128, 256, 128]` src, dim=1 | float32 | amax | False | ~1.6 ms |
| `[32, 128, 512]` input, `[32, 256, 512]` src, dim=1 | float16 | mean | True | ~1.3 ms |

---

## Actual Behavior

| Shape | Build | CPU time | Slowdown |
|-------|-------|----------|----------|
| Shape A — float32 amax | 20260428 | 1.59 ms | — |
| Shape A — float32 amax | 20260505 | 1.57 ms | — |
| Shape A — float32 amax | **20260512** | **24.00 ms** | **15×** |
| Shape B — float16 mean | 20260428 | 1.33 ms | — |
| Shape B — float16 mean | 20260505 | 1.31 ms | — |
| Shape B — float16 mean | **20260512** | **77.98 ms** | **58×** |

GPU time is ~0.14 ms across all three builds for both shapes — the regression is CPU-only.

### Versions

```
PyTorch version: 2.13.0.dev20260512+cu130
Python:          3.12.3 (GCC 13.3.0)
OS:              Linux 6.14.0-37-generic x86_64 (Ubuntu, glibc 2.39)
numpy:           2.4.4
GPU:             NVIDIA GeForce RTX 5090 (sm_120)
CUDA:            13.0
```

Bisect range: works in `2.13.0.dev20260505+cu130`, broken in `2.13.0.dev20260512+cu130`.

cc @jerryzh168 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @pragupta

**Comment by seanonymoussubmission** [comment]
Would you mind please checking this issue to see if it's a bug or not? Thanks

**Comment by malfet** [comment]
I can not repro it on my end
## Environment

- **CPU:** Intel Xeon Platinum 8559C (96 cores / 192 threads, AVX-512)
- **OS:** Linux 6.12.77 (Amazon Linux 2023, KVM)
- **Python:** 3.12.3
- **Default PyTorch threads:** 36

## Benchmark

Two shapes from the issue:

- **Shape A:** input (128,64,128) float32, src (128,256,128), reduce="amax", include_self=False
- **Shape B:** input (32,128,512) float16, src (32,256,512), reduce="mean", include_self=True

### Results: `2.13.0.dev20260505+cpu` (reported as working)

| Threads | Shape A (ms) | Shape B (ms) |
|---------|-------------|-------------|
| 1       | 37.15       | 22.77       |
| 2       | 18.67       | 13.57       |
| 4       | 9.33        | 7.04        |
| 8       | 4.68        | 4.17        |
| 16      | 2.38        | 2.68        |
| 36      | 1.13        | 3.00        |

### Results: `2.13.0.dev20260512+cu130` (reported as broken)

| Threads | Shape A (ms) | Shape B (ms) |
|---------|-------------|-------------|
| 1       | 37.20       | 22.72       |
| 2       | 18.64       | 14.12       |
| 4       | 9.34        | 4.93        |
| 8       | 4.69        | 2.50        |
| 16      | 2.40        | 1.28        |
| 36      | 1.16        | 2.15        |

### Results: `2.13.0.dev20260518+cpu` (latest nightly)

| Threads | Shape A (ms) | Shape B (ms) |
|---------|-------------|-------------|
| 1       | 36.74       | 22.74       |
| 2       | 18.42       | 14.26       |
| 4       | 9.27        | 4.78        |
| 8       | 4.65        | 4.62        |
| 16      | 2.40        | 3.13        |
| 36      | 1.13        | 3.63        |

**Comment by omid511** [comment]
can't repro here either. Ryzen 5 5500U (zen2, AVX2-only laptop, 6c/12t), cpu-only nightly 2.15.0.dev20260922. ran the issue shapes, median of 30, swept threads 1/2/4/6/12. only diff vs the issue script is I seeded with torch.manual_seed(0) instead of Generator(880014), so the values differ but shapes and timing are the same.

| threads | Shape A f32 amax (ms) | Shape B f16 mean (ms) |
|---:|---:|---:|
| 1 | 49.97 | 39.01 |
| 2 | 50.36 | 45.34 |
| 4 | 16.98 | 28.19 |
| 6 | 24.48 | 34.04 |
| 12 | 13.79 | 23.76 |

t=1 lines up with malfet's baseline above (37.15/22.77 on the Xeon vs 49.97/39.01 here on a much slower chip), and Shape A scales about 3.6x from 1 to 12 threads. the wobble at t=2 and t=6 is just boost behavior on a mobile part. nothing looks regressed on this box.

without the reporter's cpu model and thread count there isn't much to compare against. 1 thread vs default threads alone could explain the reported ratios.

<details>
<summary>script</summary>

```python
import statistics, time
import torch

def bench_cpu(fn, n=30):
    fn()  # warmup
    samples = []
    for _ in range(n):
        t0 = time.perf_counter()
        fn()
        samples.append(time.perf_counter() - t0)
    return statistics.median(samples) * 1e3  # ms

torch.manual_seed(0)
input_a = torch.randn(128, 64, 128, dtype=torch.float32)
index_a = torch.randint(0, 64, (128, 256, 128), dtype=torch.int64)
src_a = torch.randn(128, 256, 128, dtype=torch.float32)
input_b = torch.randn(32, 128, 512, dtype=torch.float16)
index_b = torch.randint(0, 128, (32, 256, 512), dtype=torch.int64)
src_b = torch.randn(32, 256, 512, dtype=torch.float16)

for t in (1, 2, 4, 6, 12):
    torch.set_num_threads(t)
    ta = bench_cpu(lambda: torch.scatter_reduce(input_a.clone(), 1, index_a, src_a, reduce="amax", include_self=False))
    tb = bench_cpu(lambda: torch.scatter_reduce(input_b.clone(), 1, index_b, src_b, reduce="mean", include_self=True))
    print(f"threads={t:>2}  ShapeA={ta:.2f} ms  ShapeB={tb:.2f} ms")
```

</details>
